File: day002_sparse_arrays/solution.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>

using namespace std;

// Function to find the matches using the Optimized Approach
vector<int> matchingStringsOptimized(const vector<string>& stringList, const vector<string>& queries) {
    // 1. Build Frequency Map (O(N))
    unordered_map<string, int> freqMap;
    for (const string& s : stringList) {
        freqMap[s]++;
    }

    // 2. Process Queries (O(M))
    vector<int> results;
    for (const string& q : queries) {
        // Look up the count in O(1) average time, defaulting to 0 if not found
        results.push_back(freqMap.count(q) ? freqMap.at(q) : 0);
    }

    return results;
}

// Function to find the matches using the Brute Force Approach
// Time Complexity: O(N * M)
// Space Complexity: O(1)
vector<int> matchingStringsBruteForce(const vector<string>& stringList, const vector<string>& queries) {
    vector<int> results;
    
    // Iterate through every query (O(M))
    for (const string& query : queries) {
        int count = 0;
        // For every query, iterate through every string in the list (O(N))
        for (const string& s : stringList) {
            if (query == s) {
                count++;
            }
        }
        results.push_back(count);
    }
    return results;
}

File: day002_sparse_arrays/solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Result {

    /**
     * Optimized Approach: Uses a HashMap for frequency counting.
     * Time Complexity: O(N + M)
     * Space Complexity: O(U) where U is the number of unique strings.
     */
    public static List<Integer> matchingStringsOptimized(List<String> stringList, List<String> queries) {
        // 1. Build Frequency Map (O(N))
        Map<String, Integer> freqMap = new HashMap<>();
        for (String s : stringList) {
            // getOrDefault is a concise way to handle the increment
            freqMap.put(s, freqMap.getOrDefault(s, 0) + 1);
        }

        // 2. Process Queries (O(M))
        List<Integer> results = new ArrayList<>();
        for (String q : queries) {
            // getOrDefault returns 0 if the key is not found (O(1) average lookup)
            results.add(freqMap.getOrDefault(q, 0));
        }

        return results;
    }

    /**
     * Brute Force Approach: Uses nested loops for matching.
     * Time Complexity: O(N * M)
     * Space Complexity: O(1)
     */
    public static List<Integer> matchingStringsBruteForce(List<String> stringList, List<String> queries) {
        List<Integer> results = new ArrayList<>();
        
        // Iterate through every query (O(M))
        for (String query : queries) {
            int count = 0;
            // For every query, iterate through every string in the list (O(N))
            for (String s : stringList) {
                if (query.equals(s)) {
                    count++;
                }
            }
            results.add(count);
        }
        return results;
    }
}